### ISSUE TITLE
Rule proposal: Consistent destructuring

### ISSUE THREAD

**Issue by futpib** [sindresorhus/eslint-plugin-unicorn#248]
[<img alt="Issuehunt badges" src="https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/248/badge.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/248)

If a variable (or a property, but probably not a call or complex expression) is destructured somewhere, there is rarely good reason to use it's properties instead of the destructured variables. 

```jsx
const {size, color} = this.props;

return <Foo size={size}>
  <Bar color={color}>
    <Buz shape={this.props.shape} />
  </Bar>
</Foo>;
```

A later editor of the code can simply miss the fact that `this.props` is destructured at the beginning of a large enough block and use the wanted value directly, leading to this inconsistency.

Alternative rule name suggestions are welcome.

## Fail
```js
const {a} = foo;
console.log(a, foo.b); 
```

```js
const {a} = foo;
console.log(foo.a); 
```

## Pass
```js
const {a} = foo;
console.log(a, foo.b()); 
```
```js
const {a, b} = foo;
console.log(a, b);
```
```js
console.log(foo.a, foo.b);
```

> There is a $80.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/248).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

**Comment by sindresorhus** [comment]
👍 Good idea.

**Comment by MrHen** [comment]
A few more variants of this:

```
const {
  a: {
    b,
  },
} = foo;
console.log(foo.a.c);
```

```
const {
  a: {
    b,
  },
} = foo;
console.log(foo.a);
```

**Comment by IssuehuntBot** [comment]
@issuehunt has funded $80.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/248) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/248).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by medusalix** [comment]
I'm working on it 😀.

**Comment by nickofthyme** [comment]
Per #759 to be included in this new rule.

> It would be nice to have a rule that combines/simplifies a duplicated object destructuring. Similar to [`no-duplicates`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-duplicates.md).
> 
> ### Bad
> ```ts
> const { defaultColor } = module;
> const { transparentColor } = module;
> const { defaultD3Color } = module;
> ```
> 
> ```ts
> const { defaultColor } = module;
> let { transparentColor } = module;
> const { defaultD3Color } = module;
> ```
> 
> ### Good
> ```ts
> const { defaultColor, transparentColor, defaultD3Color } = module;
> ```

Thanks!

**Comment by fregante** [comment]
Actually that `let` would be the exception. If `let` is used, it’s because it needs to be reassigned.

**Comment by nickofthyme** [comment]
@fregante correct, sorry idk why the rest of that got truncated from the description in #759. I updated my comment.
